sympy/angles/euler.py

```diff
@@ -0,0 +1,688 @@
+            else:
+                raise ValueError('Only commutative expressions can be '
+                                 'multiplied with an Euler.')
+
+        if lhs.seq != rhs.seq:
+            warnings.warn('Different sequences detected, '
+                          'returning sequence of left element.')
+        else:
+            warnings.warn('Euler angles are bad for composing, '
+                          'consider using quaternions instead.')
+
+        q = lhs.to_quaternion() * rhs.to_quaternion()
+        return Euler.from_quaternion(q, lhs.seq)
+
+    def set_sequence(self, new_sequence):
+        """Gets the equivalent Euler angles in the new sequence."""
+        seq = self.seq
+        if seq == new_sequence:
+            return self
+
+        q = self.to_quaternion()
+        return Euler.from_quaternion(q, new_sequence)
+
+    def switch_set(self):
+        """Gets a different set of Euler angles int the same sequence
+        that correspond to the same rotation."""
+        lamb = 0 if self.symmetric else pi
+        alpha = self.alpha + pi
+        beta = -self.beta + lamb
+        gamma = self.gamma + pi
+        return Euler(alpha, beta, gamma, self.seq)
+
+    def mul(self, other):
+        """Multiplies Euler angles.
+
+        Euler angles are NOT suited for angle composition.
+        When multiplying two Euler angles elements, both are converted to
+        quaternions, multiplied, then converted back to Euler angles.
+
+        Parameters
+        ==========
+
+        other : Euler or symbol
+            The Euler to multiply to current (self) Euler.
+
+        Returns
+        =======
+
+        Euler
+            The resultant Euler angles after multiplying self with other
+
+        Examples
+        ========
+
+        Multiplication with a number multiplies every angle by this number:
+
+        >>> from sympy import Euler, pi
+        >>> euler = Euler(pi / 4, 0, 0, 'ZYX')
+        >>> euler.mul(2)
+        Euler(pi/2,  0,  0, ZYX)
+
+        Wrapping always happens, and the angles are between [-pi and pi],
+        so the result might seem unexpected:
+
+        >>> euler.mul(4)
+        Euler((-pi),  0,  0, ZYX)
+
+        Possible to compose Euler instances. For simple cases it might work
+        well:
+        >>> euler = Euler((-pi),  0,  0, 'ZYX')
+        >>> euler.mul(euler)
+        Euler(0,  0,  0, ZYX)
+
+        Most of the time, it is not very efficient:
+
+        >>> euler = Euler(pi / 4, 0, 0, 'ZYX')
+        >>> euler.mul(euler)
+        Euler(2*atan(2*sqrt(2)*sqrt(1/2 - sqrt(2)/4)*sqrt(sqrt(2)/4 + 1/2)),  0,  0, ZYX)
+
+        By evaluating it, we can see it is equivalent to the expected answer:
+
+        >>> euler.mul(euler).evalf()
+        Euler(1.5707963267949,  0,  0, ZYX)
+        """
+        return self._generic_mul(self, _sympify(other))
+
+    def __add__(self, other):
+        return self.add(other)
+
+    def __radd__(self, other):
+        return self.add(other)
```

Comment:
It is usually better to avoid defining these methods in Expr subclasses. I'm not sure if there is a better way though.